File: 76-minimum-window-substring/minimum-window-substring.java
class Solution {
    public String minWindow(String s, String t) {
        if(t.length() > s.length()) return "";
        // Define a HashMap to store the unique chars and count
        Map<Character, Integer> uniqueChar = new HashMap<>();

        int count = t.length();

        for(int i=0; i<count; i++) {
            if(!uniqueChar.containsKey(t.charAt(i))) {
                uniqueChar.put(t.charAt(i), 1);
            } else {
                uniqueChar.put(t.charAt(i), uniqueChar.get(t.charAt(i)) + 1);
            }
        }

        int left = 0; int right = 0; int l = -1; int r = -1;
        int minLength = s.length();
        int len = 0;
        String resultStr = "";
        while(left <= right && right < s.length()) {
            // System.out.println(count + " " + s.charAt(right));

            if(uniqueChar.containsKey(s.charAt(right))) {
                uniqueChar.put(s.charAt(right), uniqueChar.get(s.charAt(right)) - 1);
                if(uniqueChar.get(s.charAt(right)) >= 0)
                    count--;
            }

            
            while(count == 0) {
                
                if(uniqueChar.containsKey(s.charAt(left))) {
                    uniqueChar.put(s.charAt(left), uniqueChar.get(s.charAt(left)) + 1);
                    if(uniqueChar.get(s.charAt(left)) > 0)
                        count++;
                }

                if(count > 0) {
                    int diff = right - left;
                    if(diff < minLength) {
                        l = left;
                        r = right+1;
                        minLength = diff;
                    }
                }

                left++;
            }

            right++;
            
        }
        // System.out.println(l + " " + r);
        if(l >=0 && r >= 0)
            resultStr = s.substring(l, r);
        
        return resultStr;
    }
}

File: 76-minimum-window-substring/minimum-window-substring.cs
public class Solution {
    public string MinWindow(string s, string t) {
        if(t.Length > s.Length) return "";
        if(s.Equals(t)) return s;
        Dictionary<char, int> tMap = new Dictionary<char, int>();
        int minLen = s.Length+1;
        int len = s.Length;
        int l = -1, r=-1;
        int count = t.Length;
        String result = "";
        for(int i=0; i<t.Length; i++) {
            if(tMap.ContainsKey(t[i])) {
                int c = tMap[t[i]];
                tMap[t[i]] = c+1;
            } else {
                tMap[t[i]] = 1;
            }
        }

        int right = 0, left = 0;

        while(right < s.Length) {
            if(tMap.ContainsKey(s[right])) {
                int c = tMap[s[right]];
                tMap[s[right]] = c-1;
                if(tMap[s[right]] >= 0) count--;
            }

            while(count == 0) {
                if(tMap.ContainsKey(s[left])) {
                    
                    int c = tMap[s[left]];
                    tMap[s[left]] = c+1;
                    if(tMap[s[left]] > 0) count++;
                }
                if(count > 0) {
                    len = (right - left) + 1;
                    if(minLen > len) {
                        minLen = len;
                        l = left;
                        r = minLen;
                    }
                }
                left++;
            }

            right++;
        }
        
        if(l >= 0 && r >= 0) {
            result = s.Substring(l, r);
        }

        return result;
    }
}